### ISSUE TITLE
Add way to manage ACLs and Quotas for Service Accounts through the User Operator

### ISSUE THREAD

**Issue by scholzj** [strimzi/strimzi-kafka-operator#10353]
With #10317 we add support for service-account-based authentication. We should also add a way to manage Simple ACLs and Quotas with KafkaUser resources and User Operator.

**Comment by scholzj** [comment]
The tricky part he is how to deal with namespaces 🤔

**Comment by sebypp** [comment]
## Background and simplified overview for OAuth auth/authz
1. authn: Principal username is extracted from token (defaults to `sub` claim). More information [here](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/).
2. authz: username obtained from [1] is matched against Principals defined via `KafkaUser` CRD and appropriate authz is applied.

## Problem statement
`KafkaUser` CRD maps the `metadata.name` to actual Kafka Principal, which imposes [restrictions](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/) on what usernames can look like. In plain English, a Kubernetes JWT token (example below) would default to its sub.
```json
{
  "aud": [
    "https://kubernetes.default.svc"
  ],
  "exp": 1731613413,
  "iat": 1700077413,
  "iss": "https://kubernetes.default.svc",
  "jti": "ea28ed49-2e11-4280-9ec5-bc3d1d84661a", 
  "kubernetes.io": {
    "namespace": "kube-system",
    "node": {
      "name": "127.0.0.1",
      "uid": "58456cb0-dd00-45ed-b797-5578fdceaced"
    },
    "pod": {
      "name": "coredns-69cbfb9798-jv9gn",
      "uid": "778a530c-b3f4-47c0-9cd5-ab018fb64f33"
    },
    "serviceaccount": {
      "name": "coredns",
      "uid": "a087d5a0-e1dd-43ec-93ac-f13d89cd13af"
    },
    "warnafter": 1700081020
  },
  "nbf": 1700077413,
  "sub": "system:serviceaccount:kube-system:coredns" # This would be the actual Kafka Principal username.
}
```

The `system:serviceaccount:kube-system:coredns` is not a valid username that we can use for the `KafkaUser` CRD resource.

## Possible solutions

1.Implement advanced parsing for `userNameClaim` using JsonPath and/or jackson-jq to something more dynamic than just extracting a claim.
2. Add a custom annotation to the `KafkaUser` CRD that will be used as Kafka Principal username instead of the `metadata.username` defined in the CRD.

**Comment by alexymantha** [comment]
To me option 2 makes a lot of sense. Even before discovering this issue, I was expecting there to be a field in the spec or an annotation to override the principal for the Kafka user.

I am open to work on this, but I'm curious which solution the maintainers would lean towards

**Comment by scholzj** [comment]
@alexymantha I think we do not lean towards any of these. It is critical to have a clear and simple relationship between the username (principal) in Kafka and the name of the KafkaUser.

So far, I think this https://github.com/strimzi/strimzi-kafka-oauth/issues/292 is what was closest to it. Other think to consider would be whether we could move forward with making the User Operator multi-namespace, which would of course make it deterministic from the `sub` itself (which right now isn't the case as it has the namespace inside).

**Comment by alexymantha** [comment]
> So far, I think this https://github.com/strimzi/strimzi-kafka-oauth/issues/292 is what was closest to it.

I agree, I've been looking into this in parallel and that would also cover my use-case.

> which would of course make it deterministic from the sub itself (which right now isn't the case as it has the namespace inside).

I'm curious what this would like like even with multi-namespace support because the name of the KafkaUser would still not be able to match sub claim due to the unsupported `:`. Would that imply native support to match Kubernetes' sub claim from the namespace + name of the `KafkaUser`

**Comment by scholzj** [comment]
> I'm curious what this would like like even with multi-namespace support because the name of the KafkaUser would still not be able to match sub claim due to the unsupported `:`. Would that imply native support to match Kubernetes' sub claim from the namespace + name of the `KafkaUser`

So, with multi-namespace User Operator you have now 2 different pieces of information associated with each `KafkaUser` as the KafkaUser and Service Account can be always in the same namespace. So user `my-user` in namespace `my-ns` links to Service Account `my-user` in the same namespace. So it allows you to easily map it to a Service account and to its principal.

With a single namespace User Operator, you cannot do it because the KafkaUser is always in the same namespace. So you do not know is user `my-user` in namespace `kafka-namespace` should match the Service account in the same namespace, or in any other namespace.
